### DISCUSSION TITLE
cannot import name 'X' from namespace 'A.B.C' (unknown location)

### DISCUSSION THREAD

**Discussion by vadimkatsman** [pythonnet/pythonnet#2116] [Q&A] [answered]
I am trying to import a class from the known and properly spelled namespace from the loaded assembly.

I am loading a bunch of assemblies belonging to the same library (they are loaded OK as I can check by the AddReference return)
`clr.AddReference("SCT.Common")
clr.AddReference("SCT.Common.35")
clr.AddReference("SCT.Common.40")
clr.AddReference("SCT.IO")
clr.AddReference("SCT.IO.35")
clr.AddReference("SCT.IO.40")`

As a test I am calling another known class:
`from SCT.Instrumentation import ContextInstrumentation as ci
appDomainInfo = ci.DumpAppDomainInfo()
print(appDomainInfo)`

Imports OK, runs OK, returns printable result.

Then I am loading another class:
`from SCT.IO.TemplateParser import TemplateBuilder as tb`

This time around I am getting an error:
`Traceback (most recent call last):
  File "D:\_WorkRoot\Tests\...\sctlib.py", line 35, in <module>
    from SCT.IO.TemplateParser import TemplateBuilder as tb
ImportError: cannot import name 'TemplateBuilder' from 'SCT.IO.TemplateParser' (unknown location)`

Context: 
0) Python 3.10 PythonNet 3.0.1
1) I am using assembly names inside AddRreference calls
2) SCT.Instrumentation is a namespace with classes defined in SCT.Common, SCT.Common.35 and SCT.Common.40 assemblies.  The specific module I am using is defined in the the first referenced assembly.
3) SCT.IO is a namespace with classes defined in all 6 assemblies.  The namespace SCT.IO.TemplateParser is defined in the very last referenced assembly ("SCT.IO.40")
4) SCT.IO.40 has a target framework as .Net Framework 4.8. But I can see successful reference:

`asmSCTIO40 = clr.AddReference("SCT.IO.40")
print(asmSCTIO40)`
"SCT.IO.40, Version=400.2023.201.25619, Culture=neutral, PublicKeyToken=null"

I noticed few discussion on StackOverflow - it seems I am following everything I have to.

**Comment by lostmsu** [comment, marked as answer]
Try calling `GetTypes` on the assembly instance. E.g. `asmSCTIO40.GetTypes()`

**Reply by vadimkatsman** [reply]
Hm ...  
System.Type[]

Blank array of System.Type objects.  Rather unexpected.

Based on my understanding, the asmSCTIO40 is of System.Assembly type and GetTypes() its method - so the collection comes from the .Net dicrectly.

If we are dealing with SCT.IO.40 being .Net 4.8 assembly while the ClrLoader is 4.7 assembly (but appdomain considers the whole thing 4.0 assembly) I can test types from assemblies compiled for .Net 2.0 and .Net 3.5.  Will update you shortly.

**Reply by vadimkatsman** [reply]
I gave a misleading report - ToString on types collection returns just a class name, but arrays are full of types:

> SCT.Common, Version=400.2023.201.25616, Culture=neutral, PublicKeyToken=null
System.Type[]
number of types = 500
SCT.IO, Version=400.2023.201.25617, Culture=neutral, PublicKeyToken=null
System.Type[]
number of types = 114
SCT.IO.35, Version=400.2023.201.25618, Culture=neutral, PublicKeyToken=null
System.Type[]
number of types = 1
SCT.IO.40, Version=400.2023.201.25619, Culture=neutral, PublicKeyToken=null
System.Type[]
number of types = 98

Looks like everything is loaded.  Will iterate for type I am actually trying to load ...
The good news, target framework does not prevent from assembly loading its classes

**Reply by vadimkatsman** [reply]
Getting into loaded types allowed to spot issues ...
Dumping assemblies and types is not something we do every day in .Net :)
Thanks!
